File: AOC_2022/Day 6/6.py
file = open("input.txt","r")

lines = file.read().splitlines()

for line in lines:
    result,l = 0,0
    s = set()
    s.add(line[0])
    for i in range(1,len(line)):
        if line[i] in s:
            while l < i:
                if line[l] != line[i]:
                    s.remove(line[l])
                    l += 1
                else:
                    l += 1
                    break            
        else:
            s.add(line[i])
            if len(s) == 4:
                result = i+1
                break
    print("The Result for the 1st star is: " + str(result))


for line in lines:
    result,l = 0,0
    s = set()
    s.add(line[0])
    for i in range(1,len(line)):
        if line[i] in s:
            while l < i:
                if line[l] != line[i]:
                    s.remove(line[l])
                    l += 1
                else:
                    l += 1
                    break            
        else:
            s.add(line[i])
            if len(s) == 14:
                result = i+1
                break
    print("The Result for the 2nd star is: " + str(result))

File: AOC_2022/Day 6/6.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstring>
#include <cctype>
#include <vector>
#include <set>
using namespace std;

int main()
{
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    string line;
    // for test cases
    // while(getline(in,line)){
    //     int result,l = 0;
    //     set<char> s;
    //     s.insert(line[0]);
    //     for(int i = 1; i < line.size();i++){
    //         if(s.find(line[i]) != s.end()){
    //             while(l < i){
    //                 if(line[l] != line[i]){
    //                     s.erase(line[l]);
    //                     l++;
    //                 }
    //                 else{
    //                     l++;
    //                     break;
    //                 }
    //             }
    //         }
    //         else{
    //             s.insert(line[i]);
    //             if(s.size() == 4){
    //                 result = i+1;
    //                 break;
    //             }
    //         }
    //     }
    //     cout<<result<<"\n";
    // }
    getline(in, line);
    //cout<<line<<"\n";
    //cout<<line.size()<<"\n";
    int result,l = 0;
    set<char> s;
    s.insert(line[0]);
    for(int i = 1; i < line.size();i++){
        if(s.find(line[i]) != s.end()){
            while(l < i){
                if(line[l] != line[i]){
                    s.erase(line[l]);
                    l++;
                }
                else{
                    l++;
                    break;
                }
            }
        }
        else{
            s.insert(line[i]);
            if(s.size() == 4){
                result = i+1;
                break;
            }
        }
    }
    cout<<result<<"\n";
    // part 2
    in.clear();
    in.seekg(0, ios::beg);
    getline(in, line);
    //cout<<line<<"\n";
    //cout<<line.size()<<"\n";
    result,l = 0;
    s.clear();
    s.insert(line[0]);
    for(int i = 1; i < line.size();i++){
        if(s.find(line[i]) != s.end()){
            while(l < i){
                if(line[l] != line[i]){
                    s.erase(line[l]);
                    l++;
                }
                else{
                    l++;
                    break;
                }
            }
        }
        else{
            s.insert(line[i]);
            if(s.size() == 14){
                result = i+1;
                break;
            }
        }
    }
    cout<<result<<"\n";
    return 0;
}